### ISSUE TITLE
Using WidgetContext in a Markdown/HTML cardYour title here

### ISSUE THREAD

**Issue by vastechpro** [thingsboard/thingsboard#10110]
**Component**

 * UI

**Description**
I would like some help with using WidgetContext in a Markdown/HTML card to query for entities like device, assets, relations, etc. 

Example code:

let $injector = ctx.$scope.$injector;
let entityRelationService = $injector.get(ctx.servicesMap.get('entityRelationService'));
let result = entityRelationService.findByFrom('22546ec0-9ade-11ee-b767-61ae9d239871');

It doesn't look like this is the correct usage. Do you have example code that shows how we can retrieve entity information using WidgetContext in a Markdown/HTML value function in a Markdown/HTML card? I couldn't find any information online or in ThingsBoard docs/tutorials.

**Environment**

 * ThingsBoard: Cloud

**Comment by vastechpro** [comment]
Thanks you for your reply. I tried with subscribe() - I see this error on screen: No enum constant org.thingsboard.server.common.data.EntityType.undefined

The ID used is the asset ID copied from the asset details screen. I also tried using the asset name returned by the datasource in data field of the function and also tried hardcoding that name.

When I print the result, it appears as [object Object]. What is this API supposed to return? I assumed it would be a JSON structure, however, it doesn't seem like that.

**Comment by devdoob** [comment]
{
                "entityType": "ASSET",
                "id": "22546ec0-9ade-11ee-b767-61ae9d239871"
}


Pass this as parameter to function... when it asks for entityId... it means an object contiang id and entityType field

**Comment by vastechpro** [comment]
I updated the code:

let $injector = ctx.$scope.$injector;
let entityRelationService = $injector.get(ctx.servicesMap.get('entityRelationService'));
let input = {
"entityType": "ASSET",
"id": "22546ec0-9ade-11ee-b767-61ae9d239871"
};
let result = entityRelationService.findByFrom(input).subscribe();
return ""+result;

The return value is [object Object]. Do you know how to parse it?

**Comment by vastechpro** [comment]
<img width="688" alt="Screenshot 2024-02-01 at 1 28 24 PM" src="https://github.com/thingsboard/thingsboard/assets/6180900/94d22c0d-4ba2-4307-9329-aa7bced75fec">

This is what I see in the console. Does this look familiar to you?

**Comment by devdoob** [comment]
entityRelationService.findByFrom(input).subscribe(response => {
        console.log(response);
});

**Comment by vastechpro** [comment]
That works! I see the response in the console and it is an Array object which I can parse. Thank you so much! One last thing: I am trying to use this response in the HTML code in the HTML value function. How do I access the contents of this response outside of the subscribe() call?

<img width="943" alt="Screenshot 2024-02-01 at 3 26 01 PM" src="https://github.com/thingsboard/thingsboard/assets/6180900/ab9f091e-ca36-4b5e-8574-ca02e2d12b14">

**Comment by devaskim** [comment]
> How do I access the contents of this response outside of the subscribe() call?

In general, `....subscribe()` call is asynchronous against a parent function (where you construct your html/markdown template), i.e. you may get a response after the parent function has been executed.

If you construct HTML template, you may set the[ id attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id), to one of your HTML elements. When you will get a response, find that element through, e.g. JQuery, `$('#someId')` and put data into it through, e.g. [html()](https://api.jquery.com/html/) function.

=====

As alternative to your `entityRelationService.findByFrom()` approach, you can create [entity alias of relation type](https://engix-ltd.github.io/docs/user-guide/ui/aliases/#relations-query) and then get those entities through `ctx.datasources` array.

**Comment by vastechpro** [comment]
I don't think I can use either of these approaches as I have to build the HTML on the fly based on the relations returned. I also have to build a hierarchy based on sub-relations. So, 
1. I cannot preassign an ID to an HTML element as the HTML does not exist yet
2. I cannot build a hierarchical alias 

How do I use the relations service to return response synchronously so I can use it to build the HTML to return?

**Comment by devaskim** [comment]
All API calls are asynchronous

**Comment by vastechpro** [comment]
I see. So is there any other way I can use the API response in the Markdown HTML function of the HTML card other than the methods you mentioned?

**Comment by devaskim** [comment]
Don't know. Honestly, you want too lot from this widget.
There is entities hierarchy widget and finally you can create your own custom widget.

**Comment by vastechpro** [comment]
Hmmm, ok. Thank you for replying. I will try custom widget. I was not able to customize the hierarchy widget much to match the UX design.

**Comment by devaskim** [comment]
Yes, the hierarchy widget is limited in some sense, but some time ago I succeeded to implement this
![Скриншот 03-02-2024 005811](https://github.com/thingsboard/thingsboard/assets/86909129/c57b74ec-c209-4a41-b405-23ac12cd3e95)

**Comment by vastechpro** [comment]
I see. That is encouraging. Did you customize these two functions:
<img width="935" alt="Screenshot 2024-02-02 at 12 39 28 PM" src="https://github.com/thingsboard/thingsboard/assets/6180900/7c80825a-a428-46ae-a7ee-506fae7b1ec5">

**Comment by devaskim** [comment]
Only text function. Icon function in my case is `return '';`, as I remember I faced with some limitation of this function,

**Comment by vastechpro** [comment]
Would you be able to send me sample code for the text function? That will help a lot. Thank you.

**Comment by devaskim** [comment]
Here is a snippet for displaying icon with text
```
const data = nodeCtx.data;
const entity = nodeCtx.entity;
const online = data['active']; // server attribute of a device

const offlineStyle = !online ? 'filter: opacity(50%);' : '';

const iconStyle = !online ? `color:black;${offlineStyle}` : 'color:green;';
return `<mat-icon class="node-icon material-icons" role="img" aria-hidden="false" style="${iconStyle}">offline_bolt</mat-icon><span style="color: orange;font-weight: 500;">${entity.label}</span>`;
```

**Comment by vastechpro** [comment]
I think I will try to build a custom widget. It is proving quite a challenge to match the UX designs by overriding the text function and other CSS.

**Comment by devdoob** [comment]
@devaskim this heirarchy widget only work with those nested/related entities which have same relation type i.e. Contains or Manages in all their nested levels...... otherwise it fails to load ... am i correct?

**Comment by devaskim** [comment]
Looking at my query, I can conclude that it supports multiple relation filters (because of `filters` field is array)
```
const query = {
    parameters: {
        rootId: entity.id.id,
        rootType: entity.id.entityType,
        direction: "FROM",
        maxLevel: 1
    },
    filters: [{
        relationType: "Contains",
        entityTypes: ['ASSET']
    }]
};
return query;
```
Besides dashboard alias of type `Relation query` support also multiple relation filters (i.e. Contains or some custom relation name). Under the hood, both alias and hierarchy widget use the same relation API.

**Comment by vastechpro** [comment]
If anyone is curious on my final resolution, I decided to use HTML Markdown widget with multiple datasource entity aliases that fetched data that I needed to build a hierarchy HTML as per the UX design. I had to add some server attributes on data ingestion/update to indicate parent-child relationship.

Closing this question.

**Comment by vastechpro** [comment]
If anyone is curious on my final resolution, I decided to use HTML Markdown widget with multiple datasource entity aliases that fetched data that I needed to build a hierarchy HTML as per the UX design. I had to add some server attributes on data ingestion/update to indicate parent-child relationship.

Closing this question.

**Comment by darioludwigswk** [comment]
@vastechpro would you be so kind and share your solution? I am stuck with a similar problem and do not quite understand the widget context functions enough. Examples are sometimes rare. Help would be much apprechieated! Thank you!

**Comment by devaskim** [comment]
@vastechpro +1
